File: agents-sdk/appointment-booking-agent/agent.js
import OpenAI from "openai";
import { createClient } from "@supabase/supabase-js";
import { fileURLToPath } from "node:url";

// Load the shared .env resolved against this file, not the working
// directory, so the agent runs the same from any cwd.
process.loadEnvFile(fileURLToPath(new URL("../.env", import.meta.url)));

const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// Server-side client. The secret key bypasses RLS, so it is loaded from
// the environment and must never reach the browser.
const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SECRET_KEY,
);

const tools = [
  {
    type: "function",
    function: {
      name: "check_appointment_availability",
      description: "Check if a time slot is available for booking",
      parameters: {
        type: "object",
        properties: {
          date: { type: "string", description: "Date in YYYY-MM-DD format" },
          time: { type: "string", description: "Time in HH:MM format" },
        },
        required: ["date", "time"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "book_appointment",
      description: "Book an appointment slot",
      parameters: {
        type: "object",
        properties: {
          name: { type: "string" },
          date: { type: "string" },
          time: { type: "string" },
          service: { type: "string" },
        },
        required: ["name", "date", "time", "service"],
      },
    },
  },
];

async function executeTool(toolName, toolInput) {
  if (toolName === "check_appointment_availability") {
    // Ask both tables at once: does the slot exist, and is it taken?
    // supabase-js resolves with { data, error } rather than throwing,
    // so the errors have to be checked explicitly.
    const [slot, booked] = await Promise.all([
      supabase
        .from("availability_slots")
        .select("id")
        .eq("date", toolInput.date)
        .eq("time", toolInput.time),
      supabase
        .from("appointments")
        .select("id")
        .eq("date", toolInput.date)
        .eq("time", toolInput.time),
    ]);
    if (slot.error) throw new Error(slot.error.message);
    if (booked.error) throw new Error(booked.error.message);

    if (slot.data.length === 0)
      return "Not offered - the business has no slot at that date and time";
    return booked.data.length === 0 ? "Available" : "Booked";
  }

  if (toolName === "book_appointment") {
    const { error } = await supabase.from("appointments").insert([toolInput]);
    if (error) throw new Error(error.message);
    return `Appointment booked for ${toolInput.name}`;
  }

  throw new Error(`Unknown tool: ${toolName}`);
}

async function runAgent(userMessage) {
  const messages = [
    // The model has no notion of the current date. Supplying it lets
    // relative dates such as "tomorrow" resolve correctly.
    {
      role: "system",
      content: `You are an appointment booking assistant. Today is ${
        new Date().toISOString().split("T")[0]
      }. Always check availability before booking.`,
    },
    { role: "user", content: userMessage },
  ];

  while (true) {
    const response = await client.chat.completions.create({
      model: "gpt-4o-mini",
      messages,
      tools,
    });

    const message = response.choices[0].message;

    // No tool calls means the model has produced its final answer.
    if (!message.tool_calls?.length) return message.content ?? "";

    messages.push(message);

    // A single turn may contain several tool calls. Each one requires its
    // own "tool" message in the reply, keyed by tool_call_id.
    for (const call of message.tool_calls) {
      let content;
      try {
        // `arguments` arrives as a JSON string, not an object.
        content = await executeTool(
          call.function.name,
          JSON.parse(call.function.arguments),
        );
      } catch (err) {
        // Report failures back to the model. Omitting a response for a
        // pending tool call causes the next request to fail with a 400.
        content = `Error: ${err.message}`;
      }
      messages.push({ role: "tool", tool_call_id: call.id, content });
    }
  }
}

// Example run
const result = await runAgent("I wanna book an appointment for a head massage today at 3:30pm. My name is John Doe.");
console.log(result);

File: agents-sdk/appointment-booking-agent-in-python/agent.py
"""Appointment booking agent, built on the OpenAI Agents SDK.

The SDK owns the agent loop: it decides when to call a tool, runs it, feeds
the result back, and repeats until the model has an answer. Compare with
../appointment-booking-agent-in-js, where that loop is written out by hand.
"""

import asyncio
import os
import datetime
from pathlib import Path

from agents import Agent, Runner, function_tool
from dotenv import load_dotenv
from supabase import Client, create_client

# Resolved against this file, not the working directory, so the agent runs
# the same from anywhere.
load_dotenv(Path(__file__).with_name(".env"))

SUPABASE_SECRET_KEY = os.environ["SUPABASE_SECRET_KEY"]

supabase: Client = create_client(os.environ["SUPABASE_URL"], SUPABASE_SECRET_KEY)


# The docstring is not decoration: the SDK turns it into the tool description
# and the per-argument schema the model sees. Type hints become the types.
@function_tool
def check_availability(date: str, time: str) -> str:
    """Check if a time slot is available for booking.

    Args:
        date: Date of the appointment in YYYY-MM-DD format.
        time: Time of the appointment in HH:MM format.
    """
    # Two questions, not one: does the business offer this slot at all, and
    # has someone already taken it? Checking only appointments would report
    # 3am on a closed day as available.
    slot = (
        supabase.table("availability_slots")
        .select("id")
        .eq("date", date)
        .eq("time", time)
        .execute()
        .data
    )
    if not slot:
        return f"The business has no slot at {date} {time}. It is not offered."

    booked = (
        supabase.table("appointments")
        .select("id")
        .eq("date", date)
        .eq("time", time)
        .execute()
        .data
    )
    if booked:
        return f"Sorry, {date} at {time} is already booked."
    return f"{date} at {time} is available for booking."


@function_tool
def book_appointment(date: str, time: str, name: str, service: str) -> str:
    """Book an appointment.

    Args:
        date: Date of the appointment in YYYY-MM-DD format.
        time: Time of the appointment in HH:MM format.
        name: Name of the person booking the appointment.
        service: Type of service requested.
    """
    row = {"date": date, "time": time, "name": name, "service": service}
    # supabase-py raises on API errors, so no error tuple to unpack here.
    # A unique constraint on (date, time) means a slot taken between the
    # availability check and this insert fails here rather than double books.
    supabase.table("appointments").insert(row).execute()
    return f"Appointment booked for {name} on {date} at {time} for {service}."


@function_tool
def list_open_slots(date: str) -> str:
    """List every free slot on a given day, so an alternative can be offered.

    Args:
        date: Date to look at, in YYYY-MM-DD format.
    """
    slots = (
        supabase.table("availability_slots")
        .select("time")
        .eq("date", date)
        .order("time")
        .execute()
        .data
    )
    if not slots:
        return f"The business is not open on {date}."

    booked = (
        supabase.table("appointments").select("time").eq("date", date).execute().data
    )
    taken = {row["time"] for row in booked}
    free = [row["time"] for row in slots if row["time"] not in taken]
    if not free:
        return f"Every slot on {date} is booked."
    return f"Open slots on {date}: {', '.join(free)}"


def build_agent() -> Agent:
    # The model has no notion of the current date, and every relative phrase
    # ("tomorrow", "next Tuesday") depends on it. Local time, not UTC: using
    # UTC books a day early whenever the local date is already ahead.
    today = datetime.date.today()
    return Agent(
        name="Appointment booking agent",
        instructions=(
            "You are an appointment booking assistant. "
            f"Today is {today.isoformat()}, a {today.strftime('%A')}. "
            "Resolve relative dates yourself before calling a tool. "
            "Always check availability before booking. If the slot is taken "
            "or not offered, call list_open_slots and suggest alternatives "
            "rather than booking something the customer did not ask for. "
            "Ask for the customer's name if they have not given one."
        ),
        tools=[check_availability, book_appointment, list_open_slots],
        model="gpt-4o-mini",
    )


async def main() -> None:
    result = await Runner.run(
        build_agent(),
        "Book me a haircut tomorrow at 9:00 AM. My name is Hamza.",
    )
    print(result.final_output)


if __name__ == "__main__":
    asyncio.run(main())
